src/providers/wfs/oapif/qgsoapiffeaturedownloaderimpl.cpp

```diff
@@ -174,10 +179,283 @@ void QgsOapifFeatureDownloaderImpl::run( bool serializeFeatures, long long maxFe
+          QgsMessageLog::logMessage( mErrorMessage, tr( "OAPIF" ) );
+        }
+        break;
+      }
+      else if ( gmlParser.isException() )
+      {
+        // Only process the exception report if we get the full error response.
+        if ( !finished )
+          continue;
+        success = false;
+
+        mErrorMessage = tr( "Server generated an exception in GetFeature response" ) + ": " + gmlParser.exceptionText();
+        QgsMessageLog::logMessage( mErrorMessage, tr( "OAPI" ) );
+
+        break;
+      }
+      // Test error code only after having let a chance to the parser to process the ExceptionReport
+      else if ( mErrorCode != NoError )
+      {
+        success = false;
+        break;
+      }
 
+      if ( totalDownloadedFeatureCount == 0 && useProgressDialog && !mResponseHeaders.empty() && !mTimer && mNumberMatched < 0 )
+      {
+        for ( const auto &headerKeyValue : mResponseHeaders )
+        {
+          if ( headerKeyValue.first.compare( QByteArray( "OGC-NumberMatched" ), Qt::CaseSensitivity::CaseInsensitive ) == 0 )
+          {
+            bool ok = false;
+            const long long val = QString::fromUtf8( headerKeyValue.second ).toLongLong( &ok );
+            if ( ok )
+            {
+              mNumberMatched = val;
+
+              CREATE_PROGRESS_TASK( QgsOapifFeatureDownloaderImpl );
+            }
+          }
+        }
+      }
+
+
+      QVector<QgsGmlStreamingParser::QgsGmlFeaturePtrGmlIdPair> featurePtrList = gmlParser.getAndStealReadyFeatures();
+
+      totalDownloadedFeatureCount += featurePtrList.size();
+
+      if ( !mStop )
+      {
+        emit updateProgress( totalDownloadedFeatureCount );
+      }
+
+      if ( featurePtrList.size() != 0 )
+      {
+        QVector<QgsFeatureUniqueIdPair> featureList;
+        for ( int i = 0; i < featurePtrList.size(); i++ )
+        {
+          QgsGmlStreamingParser::QgsGmlFeaturePtrGmlIdPair &featPair = featurePtrList[i];
+          QgsFeature &f = *( featPair.first );
+          QString gmlId( featPair.second );
+          if ( gmlId.isEmpty() )
+          {
+            // Should normally not happen on sane WFS sources, but can happen with
+            // Geomedia
+            gmlId = QgsBackgroundCachedSharedData::getMD5( f );
+            if ( !mShared->mHasWarnedAboutMissingFeatureId )
+            {
+              QgsDebugError( QStringLiteral( "Server returns features without fid/gml:id. Computing a fake one using feature attributes" ) );
+              mShared->mHasWarnedAboutMissingFeatureId = true;
+            }
+          }
+
+          // If receiving a geometry collection, but expecting a multipoint/...,
+          // then try to convert it
+          if ( f.hasGeometry() && f.geometry().wkbType() == Qgis::WkbType::GeometryCollection && ( mShared->mWKBType == Qgis::WkbType::MultiPoint || mShared->mWKBType == Qgis::WkbType::MultiLineString || mShared->mWKBType == Qgis::WkbType::MultiPolygon ) )
```

Comment:
Just wondering if you considered using the QgsVectorLayerUtils::makeFeatureCompatible method.